src/ore/src/region.rs

```diff
@@ -148,3 +150,232 @@ impl<T> LgAllocRegion<T> {
+            Region::MMap(MMapRegion { inner, handle })
+        })
+    }
+
+    /// Create a region depending on the capacity.
+    ///
+    /// The capacity of the returned region must be at least as large as the requested capacity,
+    /// but can be larger if the implementation requires it.
+    ///
+    /// Returns a [`Region::MMapRegion`] if possible, and falls back to [`Region::Heap`] otherwise.
+    #[must_use]
+    pub fn new_auto(capacity: usize) -> Region<T> {
+        match Region::new_mmap(capacity) {
+            Ok(r) => return r,
+            Err(lgalloc::AllocError::Disabled) | Err(lgalloc::AllocError::InvalidSizeClass(_)) => {}
+            Err(e) => {
+                eprintln!("lgalloc error: {e}, falling back to heap");
+            }
+        }
+        // Fall-through
+        Region::new_heap(capacity)
+    }
+
+    /// Clears the contents of the region, without dropping its elements.
+    ///
+    /// # Safety
+    ///
+    /// Discards all contends. Elements are not dropped.
+    #[inline]
+    pub unsafe fn clear(&mut self) {
+        match self {
+            Region::Heap(vec) => vec.set_len(0),
+            Region::MMap(inner) => inner.clear(),
+        }
+    }
+
+    /// Returns the capacity of the underlying allocation.
+    #[inline]
+    #[must_use]
+    pub fn capacity(&self) -> usize {
+        match self {
+            Region::Heap(vec) => vec.capacity(),
+            Region::MMap(inner) => inner.inner.capacity(),
+        }
+    }
+
+    /// Returns the number of elements in the allocation.
+    #[inline]
+    #[must_use]
+    pub fn len(&self) -> usize {
+        match self {
+            Region::Heap(vec) => vec.len(),
+            Region::MMap(inner) => inner.len(),
+        }
+    }
+
+    /// Returns true if the region does not contain any elements.
+    #[inline]
+    #[must_use]
+    pub fn is_empty(&self) -> bool {
+        match self {
+            Region::Heap(vec) => vec.is_empty(),
+            Region::MMap(inner) => inner.is_empty(),
+        }
+    }
+
+    /// Dereference to the contained vector
+    #[inline]
+    #[must_use]
+    pub fn as_vec(&self) -> &Vec<T> {
+        match self {
+            Region::Heap(vec) => vec,
+            Region::MMap(inner) => &inner.inner,
+        }
+    }
+
+    /// Extend the underlying region from the iterator.
+    ///
+    /// Care must be taken to not re-allocate the inner vector representation.
+    #[inline]
+    pub fn extend<I: IntoIterator<Item = T> + ExactSizeIterator>(&mut self, iter: I) {
+        assert!(self.capacity() - self.len() >= iter.len());
+        // SAFETY: We just asserted that we have sufficient capacity.
+        unsafe { self.as_mut_vec().extend(iter) };
+    }
+
+    /// Obtain a mutable reference to the inner vector representation.
+    ///
+    /// Unsafe because the caller has to make sure that the vector will not reallocate.
+    #[inline]
+    pub unsafe fn as_mut_vec(&mut self) -> &mut Vec<T> {
```

Comment:
as we chatted about in-person, if we could make this API private and then all callers can assert there is enough capacity, that would be great!